### PR TITLE
feat(ci): Add the linter gitlab

### PR DESCRIPTION
**Pull request by chouetz**
### What does this PR do?
Add a linter to gitlab configuration. 
```
inv linter.gitlab-ci [-t main,mq,release,all] [-c '{"CI_COMMIT_BRANCH": "main"}']
```
This will test several configuration and assess the pipeline can be generated in such conditions.
You can:
- test predefined configurations (`main`, `mq` or `release`) with the `-t` option
- test a custom config by passing either a yaml file path or a json to the `-c` command

### Motivation
Improve the stability when modifying the pipeline

### Additional Notes
- Added as pre-commit hook on modified files: duration is 1mn for 4 configs
- We can submit a custom config to test

### CODE REVIEW

File: tasks/libs/ciproviders/gitlab.py [hunk 1]
```diff
@@ -396,21 +405,141 @@ def generate_gitlab_full_configuration(input_file):
     full_configuration = {}
     for yaml_file in yaml_contents:
         full_configuration.update(yaml_file)
+    # Override some variables with a dedicated context
+    if context:
+        full_configuration["variables"].update(context)
+    if compare_to:
```
Comment (chouetz) [inline tasks/libs/ciproviders/gitlab.py:411]: This option is not used yet, I'm still trying to see if a change in the configuration can work on the gitlab API call

File: tasks/libs/ciproviders/gitlab.py [hunk 2]
```diff
@@ -396,21 +405,141 @@ def generate_gitlab_full_configuration(input_file):
...
+        ("RUN_ALL_BUILDS", ["true"]),
+        ("RUN_E2E_TESTS", ["auto"]),
+        ("RUN_KMT_TESTS", ["on"]),
+        ("RUN_UNIT_TESTS", ["on"]),
+        ("TESTING_CLEANUP", ["true"]),
+    ]
+    release_contexts = [
+        ("BUCKET_BRANCH", ["stable"]),
+        ("CI_COMMIT_BRANCH", ["7.42.x"]),
+        ("CI_COMMIT_TAG", ["3.2.1", "1.2.3-rc.4"]),
```
Comment (chouetz) [inline tasks/libs/ciproviders/gitlab.py:483]: I defined the contexts this way to have the possibility to test combinations. However as the API call is quite long (~15s) we might prefer to test only few critic combinations that can be written as `yml` templates for a better readability

File: tasks/libs/ciproviders/gitlab.py [hunk 3]
```diff
@@ -396,21 +405,141 @@ def generate_gitlab_full_configuration(input_file):
...
+        return
+    for value in contexts[0][1]:
+        context.append((contexts[0][0], value))
+        generate_contexts(contexts[1:], context, all_contexts)
+        context.pop()
+
+
+def load_context(context):
+    """
+    Load a context either from a file or from a json string
```
Comment (pducolin) [inline tasks/libs/ciproviders/gitlab.py:529]:
```suggestion
    Load a context either from a yaml file or from a json string
```

**Comment by pducolin** [comment]
This is great! Can you add the local command to the PR description?

**Review by pducolin — APPROVED** [review]

File: .gitlab/.pre/test_gitlab_configuration.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+test_gitlab_configuration:
+  stage: .pre
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/deb_x64$DATADOG_AGENT_BUILDIMAGES_SUFFIX:$DATADOG_AGENT_BUILDIMAGES
+  tags: ["arch:amd64"]
+  rules:
+    - !reference [.on_gitlab_changes]
+  script:
+    - source /root/.bashrc
+    - export GITLAB_TOKEN=$($CI_PROJECT_DIR/tools/ci/aws_ssm_get_wrapper.sh $GITLAB_SCHEDULER_TOKEN_SSM_NAME)
+    - inv linter.gitlab-ci
```
Comment (CelianR) [inline .gitlab/.pre/test_gitlab_configuration.yml:10]: **🥜 nitpick:** ‏Can you add a trailing newline ?

File: tasks/unit-tests/testdata/gitlab_main_context_template.yml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+variables:
+  BUCKET_BRANCH: 'dev'
+  CI_COMMIT_BRANCH: 'main'
+  CI_COMMIT_TAG: ''
+  CI_PIPELINE_SOURCE: 'pipeline'
+  DEPLOY_AGENT: 'true'
+  RUN_ALL_BUILDS: 'true'
+  RUN_E2E_TESTS: 'on'
+  RUN_KMT_TESTS: 'on'
+  RUN_UNIT_TESTS: 'true'
```
Comment (CelianR) [inline tasks/unit-tests/testdata/gitlab_main_context_template.yml:11]: **🥜 nitpick:** ‏Can you add a line feed ?

File: tasks/unit-tests/testdata/gitlab_mq_context_template.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+---
+variables:
+  BUCKET_BRANCH: 'dev'
+  CI_COMMIT_BRANCH: 'mq-working-branch-main'
+  CI_PIPELINE_SOURCE: 'pipeline'
+  DEPLOY_AGENT: 'false'
+  RUN_ALL_BUILDS: 'false'
+  RUN_E2E_TESTS: 'auto'
+  RUN_KMT_TESTS: 'off'
+  RUN_UNIT_TESTS: 'off'
```
Comment (CelianR) [inline tasks/unit-tests/testdata/gitlab_mq_context_template.yml:10]: **🥜 nitpick:** ‏Can you add a line feed ?

File: tasks/linter.py [hunk 7]
```diff
@@ -351,3 +358,27 @@ def is_get_parameter_call(file):
...
+    for context in all_contexts:
+        print("Test gitlab configuration with context: ", context)
+        config = generate_gitlab_full_configuration(".gitlab-ci.yml", dict(context))
+        gitlab = Gitlab(api_token=get_gitlab_token())
+        res = gitlab.lint(config)
+        print(f"Config is valid: {res['valid']}")
+        if len(res["warnings"]) > 0:
+            print(f"Warnings: {res['warnings']}")
+        if not res["valid"]:
+            print(f"Errors: {res['errors']}")
```
Comment (CelianR) [inline tasks/linter.py:383]: Should this be printed in red / on stderr / using a logger ? (same for warnings)

**Review by CelianR — APPROVED** [review]

File: tasks/linter.py (hunk 7 above)
Comment (chouetz) [reply tasks/linter.py:383]: Used the current solution (`color_message`). Will involve a logger later

**Pull request merged**